### PR TITLE
Improve streams ReadPendingScope cleanup

### PR DESCRIPTION
**Pull request by jasnell**
Changes the `ReadPendingScope` from a struct to a function that includes (hopefully) appropriate error handling logic.

### CODE REVIEW

File: src/workerd/api/streams/standard.c++ [hunk 1]
```diff
@@ -524,22 +524,26 @@ int getHighWaterMark(const UnderlyingSource& underlyingSource,
...
-// handles this for us. It is constructed when we start a read and destructed
-// synchronously when the call to read completes.
-template <typename C>
-struct ReadPendingScope {
-  jsg::Lock& js;
-  C& controller;
-  ReadPendingScope(jsg::Lock& js, C& controller)
-      : js(js), controller(controller) {
+// close/error state change until the read call is complete. readPendingScope
+// handles this for us.
```
Comment (mikea) [inline src/workerd/api/streams/standard.c++:528]: this is a tricky code, I would very much like this sentence to continue, e.g.: `handles this for us by making sure that pendingReadCount is incremented until the read operation completes and deferring a state change until it is 0`. (did I get the code right btw?)

File: src/workerd/api/streams/standard.c++ [hunk 2]
```diff
@@ -2467,12 +2482,14 @@ kj::Maybe<jsg::Promise<ReadResult>> ReadableStreamJsController::read(
       // The ReadableStreamDefaultController does not support ByobOptions.
       // It should never happen, but let's make sure.
       KJ_ASSERT(maybeByobOptions == kj::none);
-      ReadPendingScope readPendingScope(js, *this);
-      return consumer->read(js);
+      return readPendingScope(js, *this, [&]() mutable {
```
Comment (mikea) [inline src/workerd/api/streams/standard.c++:2485]: maybe a better name is possible? how about `ensureControllerAlive` or something like that?

**Review by mikea — APPROVED** [review]
LG but would really like expanded comment and slightly better function name if possible. thanks.

File: src/workerd/api/streams/standard.c++ [hunk 3]
```diff
@@ -524,22 +524,26 @@ int getHighWaterMark(const UnderlyingSource& underlyingSource,
...
-template <typename C>
-struct ReadPendingScope {
-  jsg::Lock& js;
-  C& controller;
-  ReadPendingScope(jsg::Lock& js, C& controller)
-      : js(js), controller(controller) {
+// close/error state change until the read call is complete. readPendingScope
+// handles this for us.
+template <typename Controller>
+jsg::Promise<ReadResult> readPendingScope(
```
Comment (mikea) [inline src/workerd/api/streams/standard.c++:530]: Also, can't this be `jsg::Promise<T> ensureAlive(..., <Promise<T>()> callback)`?

**Review by hoodmane — COMMENTED** [review]
So is the point of this that previously if `readable->read` threw a C++ error, it wouldn't be caught whereas now it gets caught and used to reject the promise?

Looks reasonable to me, but I'm not great at reviewing C++ patches.

File: src/workerd/api/streams/standard.c++ (hunk 1 above)
Comment (jasnell) [inline src/workerd/api/streams/standard.c++:528]:
```suggestion
// handles this for us by making sure that pendingReadCount is incremented until the read
// operation completes and deferring a state change until it is 0
```

File: src/workerd/api/streams/standard.c++ (hunk 2 above)
Comment (harrishancock) [reply src/workerd/api/streams/standard.c++:2487]: `runDeferringCloseAndErrorStateChanges()`? :)

**Review by harrishancock — APPROVED** [review]

File: src/workerd/api/streams/standard.c++ (hunk 3 above)
Comment (jasnell) [reply src/workerd/api/streams/standard.c++]: Perhaps, but I think keeping this as is should be fine.

**Pull request merged**